File: backend/app/utils/phone.py
"""International mobile number normalization, validation, and formatting."""

import re

# Supported corridor / onboarding country calling codes (E.164)
SUPPORTED_COUNTRY_CODES = ("27", "233", "263", "254", "256", "260", "234")

_E164_RE = re.compile(r"^\+[1-9]\d{7,14}$")


def normalize_phone_number(raw: str, default_country_code: str = "27") -> str:
    """Normalize to E.164 (+{country}{national_number})."""
    if not raw or not str(raw).strip():
        raise ValueError("Mobile number is required")

    cleaned = re.sub(r"[^\d+]", "", str(raw).strip())
    if cleaned.startswith("00"):
        cleaned = "+" + cleaned[2:]
    if cleaned.startswith("+"):
        digits = cleaned[1:]
    elif cleaned.startswith("0"):
        digits = default_country_code + cleaned[1:]
    else:
        digits = cleaned

    if not digits.isdigit():
        raise ValueError("Mobile number must contain only digits")

    for code in sorted(SUPPORTED_COUNTRY_CODES, key=len, reverse=True):
        if digits.startswith(code) and len(digits) >= len(code) + 7:
            return f"+{digits}"

    if len(digits) >= 9:
        return f"+{default_country_code}{digits}" if not any(
            digits.startswith(c) for c in SUPPORTED_COUNTRY_CODES
        ) else f"+{digits}"

    raise ValueError("Mobile number is too short")


def validate_phone_number(raw: str, default_country_code: str = "27") -> str:
    """Normalize and validate; returns E.164 string."""
    normalized = normalize_phone_number(raw, default_country_code)
    if not _E164_RE.match(normalized):
        raise ValueError("Invalid mobile number format")
    national = normalized[1:]
    if not any(national.startswith(code) for code in SUPPORTED_COUNTRY_CODES):
        raise ValueError(
            "Unsupported country code. Supported: South Africa (+27), Ghana (+233), "
            "Zimbabwe (+263), Kenya (+254), Uganda (+256), Zambia (+260), Nigeria (+234)"
        )
    return normalized


def format_phone_number(e164: str) -> str:
    """Human-readable display for E.164 numbers."""
    if not e164:
        return ""
    normalized = e164 if e164.startswith("+") else f"+{e164}"
    for code in sorted(SUPPORTED_COUNTRY_CODES, key=len, reverse=True):
        if normalized[1:].startswith(code):
            rest = normalized[1 + len(code) :]
            if len(rest) <= 3:
                return f"+{code} {rest}"
            return f"+{code} {rest[:3]} {rest[3:]}"
    return normalized


def is_email_identifier(value: str) -> bool:
    return "@" in value.strip()

File: frontend/src/lib/phone.ts
export function normalizePhoneNumber(raw: string, defaultCountryCode = "27"): string {
  const cleaned = raw.trim().replace(/[^\d+]/g, "");
  let digits = cleaned;
  if (digits.startsWith("00")) digits = digits.slice(2);
  if (digits.startsWith("+")) digits = digits.slice(1);
  else if (digits.startsWith("0")) digits = defaultCountryCode + digits.slice(1);

  const codes = ["27", "233", "263", "254", "256", "260", "234"];
  for (const code of codes.sort((a, b) => b.length - a.length)) {
    if (digits.startsWith(code) && digits.length >= code.length + 7) {
      return `+${digits}`;
    }
  }
  if (digits.length >= 9) return `+${defaultCountryCode}${digits}`;
  throw new Error("Invalid mobile number");
}

export function formatPhoneNumber(e164: string): string {
  if (!e164) return "";
  const normalized = e164.startsWith("+") ? e164 : `+${e164}`;
  const codes = ["27", "233", "263", "254", "256", "260", "234"];
  for (const code of codes.sort((a, b) => b.length - a.length)) {
    if (normalized.slice(1).startsWith(code)) {
      const rest = normalized.slice(1 + code.length);
      return rest.length <= 3 ? `+${code} ${rest}` : `+${code} ${rest.slice(0, 3)} ${rest.slice(3)}`;
    }
  }
  return normalized;
}

export function isEmailIdentifier(value: string): boolean {
  return value.includes("@");
}
